ast/visit.go

```diff
@@ -337,7 +337,9 @@ func (vis *GenericVisitor) Walk(x interface{}) {
 	case *object:
 		x.Foreach(func(k, v *Term) {
 			vis.Walk(k)
-			vis.Walk(x.Get(k))
+			if v := x.Get(k); v != nil {
+				vis.Walk(v)
+			}
```

Comment:
I'm not sure we want to make this change. If the lookup fails, the object must have been corrupted because the code is walking over the key-value pairs in the object--so lookup of a key should not fail.

We ran into a number of related issues in the past that were caught through fuzzing. Most of them were fixed in v0.24.0: https://github.com/open-policy-agent/opa/releases/tag/v0.24.0.

I suspect that in this case, what's happening is:

1. The object starts off as `{{1 | 0 == 0}: 2}`
2. Rewriting changes `0 == 0` into `0 = 0`
3. The *ast.Term containing `{1 | 0 == 0}` is not reallocated; instead the ast.Term#Value field is updated
4. The object lookup fails because the object's hashtable is using the hash for the original expression

Anytime we mutate values that are stored inside objects or sets we risk corrupting those collections.

One solution is to just copy objects and sets during rewriting. It's more expensive but it prevents corruption.

Reply:
Thanks, that explains what I was seeing when printf'ing my way through this. I'll give it another go! :)

Reply:
Switched to using `NewGenericTransformer`. WDYT?